oslabs-beta/apollo11
# Basic snapshot tests

**Feature**
Tests

Description
- This PR will add some basic snapshot tests using `Jest` and `Enzyme`
- Currently the main `App`, `ApolloTab` (`Event Log`), `MainDrawer` and `Performance` tabs have these basic tests
- `GraphiQL` doesn't have a direct test but is rendered since it's the default panel
- Added mocks for Chrome APIs, styled components and ThemeProvider in `__mocks__` folder
- Added a mock for `document.createRange` directly in `MainDrawer` for now
- Moved the non-Jest tests from a `__tests__` folder to an `other` folder, so that `Jest` doesn't attempt to run them
- Added `test` and `validate` scripts to `package.json`
- Excluded `__mocks__` and `__tests__` in `webpack.config.js` rules

**Installation**
- Run `npm install` to add these new packages
`babel-jest`, `enzyme-to-json`, `jest-chrome`

**How to test**
- Run `npm run test` to execute all tests
- Make a change to the UI for the components above and re-run the tests
- The snapshots should fail as there now should be differences listed in the test report
- Undo the changes and re-run the tests to ensure they pass again

If the changes are expected, update the existing snapshots by running
`jest -u`

Current test results:
```
 PASS  src/app/Events_Tab/__tests__/ApolloTab_test.js
  snapshot tests
    ✓ should mount the app (134 ms)

 PASS  src/app/Performance_Tab/__tests__/Performance_test.js
  snapshot tests
    ✓ should mount the app (211 ms)

 PASS  src/app/Panel/__tests__/app_test.js
  snapshot tests
    ✓ should mount the app (200 ms)

 PASS  src/app/Panel/__tests__/MainDrawer_test.js
  snapshot tests
    ✓ should mount the app (495 ms)

--------------------------------|---------|----------|---------|---------|------------------------------------------------
File                            | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s                              
--------------------------------|---------|----------|---------|---------|------------------------------------------------
All files                       |   32.79 |    11.71 |   33.94 |   33.28 |                                                
 Events_Tab                     |   43.75 |    10.26 |   47.06 |   41.86 |                                                
  ApolloTab.tsx                 |   81.25 |      100 |    62.5 |   78.57 | 73,77,82                                       
  Cache.tsx                     |    9.09 |        0 |       0 |      10 | 17-59                                          
  CacheDetails.tsx              |      20 |        0 |       0 |      25 | 8-12                                           
  EventDetails.tsx              |   42.86 |       25 |     100 |   33.33 | 11-18                                          
  EventLog.tsx                  |      25 |     6.67 |   33.33 |      25 | 20-47                                          
  EventPanel.tsx                |     100 |      100 |     100 |     100 |                                                
 GraphiQL_Tab                   |   55.56 |       65 |    37.5 |   55.56 |                                                
  GraphiQLPage.tsx              |   73.91 |      100 |    37.5 |   73.91 | 61,65-66,86-106                                
  GraphiqlPlugin.jsx            |   46.94 |    41.67 |    37.5 |   46.94 | 40,63,75,80,87-125,130,135-138,142,148,164-178 
 Panel                          |   80.39 |       55 |   76.47 |   80.39 |                                                
  MainDrawer.tsx                |   62.96 |       55 |      50 |   62.96 | 120-124,129,133,146-150,213                    
  app.tsx                       |     100 |      100 |     100 |     100 |                                                
 Panel/themes                   |    87.5 |       75 |      75 |    87.5 |                                                
  ThemeProvider.tsx             |   83.33 |       75 |   66.67 |   83.33 | 31-32                                          
  base.ts                       |     100 |      100 |     100 |     100 |                                                
  dark.ts                       |     100 |      100 |     100 |     100 |                                                
  normal.ts                     |     100 |      100 |     100 |     100 |                                                
 Performance_Tab                |   42.99 |     6.58 |   33.33 |   42.86 |                                                
  ArrowChip.tsx                 |     100 |      100 |     100 |     100 |                                                
  Performance.tsx               |   48.48 |        0 |   55.56 |   46.88 | 89-146,166-185                                 
  PerformanceResponsiveTest.tsx |   16.22 |        0 |       0 |   16.67 | 63,100-239                                     
  TracingDetails.tsx            |   57.14 |    17.86 |      25 |   57.14 | 98-102,109-113,120-124,178-258                 
  progressBar.tsx               |      75 |        0 |      50 |      75 | 6                                              
 utils                          |   17.05 |     3.03 |   16.67 |   17.58 |                                                
  helper.ts                     |   15.79 |        0 |       0 |   16.67 | 20-49,85-88,108,122-166                        
  messaging.ts                  |    4.55 |        0 |   33.33 |    4.55 | 15-142,154-163                                 
  networking.ts                 |     3.7 |        0 |   16.67 |    3.77 | 6-62,74-146                                    
  performanceMetricsCalcs.tsx   |   58.33 |       25 |   33.33 |   63.64 | 14-19                                          
  tracingTimeFormating.tsx      |   34.04 |     12.5 |   22.22 |   38.46 | 2-9,13-20,24-31,37-41,57-58                    
 utils/managedlog               |   14.81 |        0 |      30 |   14.81 |                                                
  eventObject.ts                |   14.81 |        0 |      30 |   14.81 | 34-148,159                                     
 utils/managedlog/lib           |    4.31 |        0 |    5.26 |    4.55 |                                                
  eventLogData.ts               |    2.88 |        0 |    6.67 |    3.06 | 19-185                                         
  eventLogNode.ts               |       0 |      100 |       0 |       0 | 11-13                                          
  objectDifference.ts           |   22.22 |        0 |       0 |   22.22 | 4,9-18                                         
--------------------------------|---------|----------|---------|---------|------------------------------------------------
Test Suites: 4 passed, 4 total
Tests:       4 passed, 4 total
Snapshots:   4 passed, 4 total
Time:        3.561 s
```
